torch/autograd/_functions/stochastic.py

```diff
@@ -1,92 +1,38 @@
-from ..stochastic_function import StochasticFunction
+import torch
+from ..function import Function
 
-# Gradient formulas are based on Simple Statistical Gradient-Following
-# Algorithms for Connectionist Reinforcement Learning, available at
-# http://incompleteideas.net/sutton/williams-92.pdf
 
-
-class Multinomial(StochasticFunction):
-
-    def __init__(self, num_samples, with_replacement):
-        super(Multinomial, self).__init__()
-        self.num_samples = num_samples
-        self.with_replacement = with_replacement
-
-    def forward(self, probs):
-        samples = probs.multinomial(self.num_samples, self.with_replacement)
-        self.save_for_backward(probs, samples)
-        self.mark_non_differentiable(samples)
+class Multinomial(Function):
+    @staticmethod
+    def forward(ctx, probs, num_samples, with_replacement):
+        samples = probs.multinomial(num_samples, with_replacement)
+        ctx.mark_non_differentiable(samples)
         return samples
 
-    def backward(self, reward):
-        probs, samples = self.saved_tensors
-        if probs.dim() == 1:
-            probs = probs.unsqueeze(0)
-            samples = samples.unsqueeze(0)
-            reward = reward.unsqueeze(0)
-        # normalize probs (multinomial accepts weights)
-        probs /= probs.sum(1, True).expand_as(probs)
-        grad_probs = probs.new().resize_as_(probs).zero_()
-        output_probs = probs.gather(1, samples)
-        output_probs.add_(1e-6).reciprocal_()
-        output_probs.neg_().mul_(reward)
-        # TODO: add batched index_add
-        for i in range(probs.size(0)):
-            grad_probs[i].index_add_(0, samples[i], output_probs[i])
-        return grad_probs
-
+    @staticmethod
+    def backward(ctx, grad_output):
+        return None, None, None
```

Comment:
We could change this to raise an error once we start detaching Variables with `requires_grad=True`. In fact you could include that in this commit

Reply:
They backward() functions actually get called for some weird reason, which is why I included the definitions.

Reply:
Maybe we should figure out why?

Reply:
I think it's because for `z = f(x, y)` if:

  x.requires_grad=False
  y.requires_grad=True

and both have `grad_fn`s defined, then z.grad_fn.next_function will include both.

I'm not particularly concerned right now for two reasons:

1. These particular functions won't last very long. Since they're simple wrappers around the torch functions they can easily be moved to C++
2. A big part of this change is to make it so that functions which don't require_grad don't have grad_fn